runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -11924,6 +11930,157 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+
+   generateRegRegInstruction(TR::InstOpCode::PUNPCKLBWRegReg, node, xmmHighReg, zeroReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(destBufferReg, destOffsetReg, 0, headerOffsetConst, cg), xmmHighReg, cg);
+
+   // increase src offset by size of imm register
+   // increase dest offset by double, to account for the byte->char inflation
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, srcOffsetReg, vectorLengthConst, cg);
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, destOffsetReg, 2 * vectorLengthConst, cg);
+
+   // LOOP BACK
+   generateLabelInstruction(TR::InstOpCode::JMP4, node, startLoop, cg);
+   generateLabelInstruction(TR::InstOpCode::label, node, endLoop, cg);
+
+   // AND length with 15 to compute residual remainder
+   // then copy and interleave 8 bytes from src buffer with 0s into dest buffer if possible
+   generateRegImmInstruction(TR::InstOpCode::AND4RegImm4, node, lengthReg, vectorLengthConst - 1, cg);
+
+   generateRegImmInstruction(TR::InstOpCode::CMP4RegImm4, node, lengthReg, 8, cg);
+   generateLabelInstruction(TR::InstOpCode::JL1, node, afterCopy8, cg);
+
+   generateRegMemInstruction(TR::InstOpCode::MOVQRegMem, node, xmmLowReg, generateX86MemoryReference(srcBufferReg, srcOffsetReg, 0, headerOffsetConst, cg), cg);
+   generateRegRegInstruction(TR::InstOpCode::PUNPCKLBWRegReg, node, xmmLowReg, zeroReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVDQUMemReg, node, generateX86MemoryReference(destBufferReg, destOffsetReg, 0, headerOffsetConst, cg), xmmLowReg, cg);
+   generateRegImmInstruction(TR::InstOpCode::SUB4RegImm4, node, lengthReg, 8, cg);
+
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, srcOffsetReg, 8, cg);
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, destOffsetReg, 16, cg);
+
+   generateLabelInstruction(TR::InstOpCode::label, node, afterCopy8, cg);
+
+   // handle residual (< 8 bytes left) & jump to copy instructions based on the number of bytes left
+   // calculate how many bytes to skip based on length;
+
+   const int copy_instruction_size = 0x9;
+
+   // size of MOVZXReg2Mem1 + S2MemReg in bytes
+   // since copy_instruction_size could change depending on which registers are allocated to scratch, srcOffsetReg and destOffsetReg
+   // we reserve them to be eax, ecx, edx, respectively
+
+   generateRegRegImmInstruction(TR::InstOpCode::IMUL4RegRegImm4, node, lengthReg, lengthReg, -copy_instruction_size, cg);
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, lengthReg, copy_instruction_size * 7, cg);
+
+   bool is64bit = cg->comp()->target().is64Bit();
+   // calculate address to jump too
+   generateRegMemInstruction(TR::InstOpCode::LEARegMem(is64bit), node, scratch, generateX86MemoryReference(duffDevice, cg), cg);
```

Comment:
This won't work without a relocation for AOT compiles (because the "duffDevice" address is only valid for this method, not if it is AOT loaded later).

Have a look at how the jump table in the tableEvaluator is implemented:

https://github.com/eclipse/omr/blob/d3053f922dd8f81535ea3396fbab3e8a51ac6e83/compiler/x/codegen/ControlFlowEvaluator.cpp#L468

Reply:
Are you sure? The code generated is rip-relative,

`48 8d 05 0e 00 00 00 // lea    rax,[rip+0xe] `

Unless this is not guaranteed...

Reply:
I wouldn't say "guaranteed", but I'd say it is extremely likely to produce RIP-relative code.  I'll double check the mem ref code.

Note that RIP-relative addressing is not available on 32-bit, so if this is something that will run on JDK8 then you'll need another solution (like the table evaluator)

Reply:
This StringLatin1.inflate does not exist of JDK8

Reply:
Maybe add a fatal assert at the beginning of this function then guarding against a 32-bit invocation.

Reply:
See [force-push](https://github.com/eclipse-openj9/openj9/compare/f08e7ba55d1e2570b9d4401f8adca3e9edc9d0b4..42b99387f681bf1b6645f32d38baf1dea925d2a8)